File: C++/283_moveZeroes.cpp
/**
* Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.
* 
* Note that you must do this in-place without making a copy of the array.
*/

// My Best solution, playing by the rules.
// Have a slow and a fast pointer. slow pointer stays at first 0 it finds, fast pointer runs to the first nonzero element.
// Swap and continue. 
// perf fluctuates 75-65% mem fluctuates 98-75%

class Solution {
public:
    void moveZeroes(vector<int>& nums) {

        int slowIndex = 0;
        int len = nums.size();
        while (nums[slowIndex] != 0)
        {
            slowIndex++;

            if (slowIndex == len) // no zeroes whatsoever
                return;
        }

        int fastIndex = slowIndex + 1;
        for (fastIndex; fastIndex < len; fastIndex++)
        {
            if (nums[fastIndex] != 0)
            {
                nums[slowIndex] = nums[fastIndex];
                nums[fastIndex] = 0;
                slowIndex++;
            }
        }
    }
};

// My initial solution, breaking the rules. make a zeroed array the same size as "nums", loop over nums and throw non zeroes into the helper array.
// Memcpy the helper array onto "nums".
// Good performance but relatively bad memory consumption.
// perf fluctuates 91-72% mem 8%

// class Solution {
// public:
//     void moveZeroes(vector<int>& nums) {
// 
//         int* newArray = new int[nums.size()] {};
//         int count = 0;
// 
//         for (int& n : nums)
//             if (n != 0)
//             {
//                 newArray[count] = n;
//                 count++;
//             }
// 
//         memcpy(nums.data(), newArray, nums.size() * sizeof(int));
//         delete[] newArray;
//     }
// };

File: Rust/283_moveZeroes.rs
/**
 * Given an integer array nums, move all 0's to the end of it while
 * maintaining the relative order of the non-zero elements.
 * 
 * Note that you must do this in-place without making a copy of the array.
 * 
 * Example 1:
 * Input: nums = [0,1,0,3,12]
 * Output: [1,3,12,0,0]
 * 
 * Example 2:
 * Input: nums = [0]
 * Output: [0]
*/

pub fn move_zeroes(nums: &mut Vec<i32>) {
    let mut zeros_to_add = 0;
    let mut i = 0;

    while i < nums.len() {
        if nums[i] == 0 {
            nums.remove(i);
            zeros_to_add += 1;
        }
        else {
            i += 1;
        }
    }

    for _i in 0..zeros_to_add {
        nums.push(0);
    }
}

/**
 * Best Solution
 */
pub fn move_zeroes(nums: &mut Vec<i32>) {
    let mut z = Vec::new();
    nums.retain(|x| {
     if *x != 0 { true } else { z.push(0); false }
    });
    &nums.append(&mut z);
}